### PR TITLE
Run kvm_unit_test on host for nested test

### PR DESCRIPTION
**Pull request by qcheng-redhat**
Add kvm_unit_test on host for nested test

ID: 1882626

Signed-off-by: Qinghua Cheng <qcheng@redhat.com>

### CODE REVIEW

**Comment by qcheng-redhat** [comment]
@PaulYuuu @coli, please help review this patch. Thanks!

File: qemu/tests/nested_kvm_unit_test.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+def run(test, params, env):
+    """
+    Run nested relstead tests in kvm-unit-test test suite
+
+    1) Clone kvm-unit-tests test suite from repo
+    2) Compile test suite
+    3) Run nested related test cases on host
+    """
+
+    kvm_unit_test_dir = "/tmp/kvm_unit_test_dir"
```
Comment (PaulYuuu) [inline qemu/tests/nested_kvm_unit_test.py:18]: duplicate with cfg.

File: qemu/tests/nested_kvm_unit_test.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    mkdir_cmd = "mkdir -p %s" % kvm_unit_test_dir
+    rmdir_cmd = "rm -rf %s" % kvm_unit_test_dir
```
Comment (PaulYuuu) [inline qemu/tests/nested_kvm_unit_test.py:20]: As above, recommend using `test.workdir`

File: qemu/tests/nested_kvm_unit_test.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    if os.path.exists(kvm_unit_test_dir):
+        process.system(rmdir_cmd)
+    process.system(mkdir_cmd)
+
+    clone_cmd = params["clone_cmd"] % kvm_unit_test_dir
+    process.system(clone_cmd, shell=True)
+    compile_cmd = params["compile_cmd"] % kvm_unit_test_dir
+    process.system(compile_cmd, shell=True)
```
Comment (PaulYuuu) [inline qemu/tests/nested_kvm_unit_test.py:29]: Is it works without `shell=True`? If yes, better to remove it.

File: qemu/tests/nested_kvm_unit_test.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    error_context.context("Run kvm_unit_tests nested tests", logging.info)
+    timeout = float(params.get("kvm_unit_test_timeout", 60))
```
Comment (PaulYuuu) [inline qemu/tests/nested_kvm_unit_test.py:32]: `params.get_numeric()` would be good.

File: qemu/tests/nested_kvm_unit_test.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    run_cmd = params["test_cmd"] % kvm_unit_test_dir
+    logging.info("Run command %s " % run_cmd)
```
Comment (PaulYuuu) [inline qemu/tests/nested_kvm_unit_test.py:34]:
```suggestion
    logging.info("Run command %s ", run_cmd)
```

File: qemu/tests/cfg/nested_kvm_unit_test.cfg [hunk 6]
```diff
@@ -0,0 +1,13 @@
+- nested_kvm_unit_test:
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_kvm_unit_test.cfg]: Renamed to kvm_unit_test_nested

File: qemu/tests/cfg/nested_kvm_unit_test.cfg [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    type = nested_kvm_unit_test
+    start_vm = no
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_kvm_unit_test.cfg]: Added.

File: qemu/tests/cfg/nested_kvm_unit_test.cfg [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    kvm_unit_test_dir = "/tmp/kvm_unit_tests"
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_kvm_unit_test.cfg]: Changed to test.workdir. 

The purpose to use /tmp/kvm_unit_tests dir is to keep the test suite after test run, re-run if there are any failure test cases.

File: qemu/tests/cfg/nested_kvm_unit_test.cfg [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    clone_cmd = "git clone --depth 1 -q https://gitlab.com/kvm-unit-tests/kvm-unit-tests.git %s"
+    compile_cmd += "cd %s && ./configure && make"
+    HostCpuVendor.amd:
+        test_cmd = "cd %s && ./x86/run ./x86/svm.flat -cpu host,+svm"
+    HostCpuVendor.intel:
+        test_cmd = "cd %s && ./x86/run ./x86/vmx.flat -cpu host"
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_kvm_unit_test.cfg]: +vmx is optional here. I will add it.

File: qemu/tests/nested_kvm_unit_test.py (hunk 1 above)
Comment (qcheng-redhat) [reply qemu/tests/nested_kvm_unit_test.py]: Removed.

File: qemu/tests/nested_kvm_unit_test.py (hunk 2 above)
Comment (qcheng-redhat) [reply qemu/tests/nested_kvm_unit_test.py]: Removed.

File: qemu/tests/nested_kvm_unit_test.py (hunk 3 above)
Comment (qcheng-redhat) [reply qemu/tests/nested_kvm_unit_test.py]: Removed for line 27 and keep for line 29.

File: qemu/tests/nested_kvm_unit_test.py (hunk 5 above)
Comment (qcheng-redhat) [reply qemu/tests/nested_kvm_unit_test.py]: Fixed.

File: qemu/tests/nested_kvm_unit_test.py (hunk 4 above)
Comment (qcheng-redhat) [reply qemu/tests/nested_kvm_unit_test.py]: Fixed.

File: qemu/tests/cfg/nested_kvm_unit_test.cfg (hunk 8 above)
[the remark this answers is not in the document]
Comment (PaulYuuu) [reply qemu/tests/cfg/nested_kvm_unit_test.cfg]: If so, you can use `test.logdir`.

[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_kvm_unit_test.cfg]: Ok, changed to test.logdir. Thanks.

File: qemu/tests/cfg/kvm_unit_test_nested.cfg [hunk 10]
```diff
@@ -0,0 +1,13 @@
+- kvm_unit_test_nested:
+    type = kvm_unit_test_nested
+    start_vm = no
+    vms = ""
```
Comment (PaulYuuu) [inline qemu/tests/cfg/kvm_unit_test_nested.cfg:4]: BTW, I see the manual case requested to boot L1 and L2 guest in the host, but this patch could not cover them. Is it expected?

File: qemu/tests/cfg/kvm_unit_test_nested.cfg [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    clone_cmd = "git clone --depth 1 -q https://gitlab.com/kvm-unit-tests/kvm-unit-tests.git %s"
+    compile_cmd += "cd %s && ./configure && make"
```
Comment (PaulYuuu) [inline qemu/tests/cfg/kvm_unit_test_nested.cfg:6]:
```suggestion
    compile_cmd = "cd %s && ./configure && make"
```

File: qemu/tests/kvm_unit_test_nested.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    Run nested relstead tests in kvm-unit-test test suite
+
+    1) Clone kvm-unit-tests test suite from repo
+    2) Compile test suite
+    3) Run nested related test cases on host
+    """
+
+    kvm_unit_test_dir = os.path.join(test.logdir, "kvm_unit_tests/")
+    logging.info("kvm_unit_test_dir %s:" % kvm_unit_test_dir)
```
Comment (PaulYuuu) [inline qemu/tests/kvm_unit_test_nested.py:19]:
```suggestion
    logging.info("kvm_unit_test_dir: %s", kvm_unit_test_dir)
```

File: qemu/tests/kvm_unit_test_nested.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    compile_cmd = params["compile_cmd"] % kvm_unit_test_dir
+    process.system(compile_cmd, shell=True)
+
+    error_context.context("Run kvm_unit_tests on host", logging.info)
+    timeout = params.get_numeric("kvm_unit_test_timeout", 60)
+    run_cmd = params["test_cmd"] % kvm_unit_test_dir
+    logging.info("Run command %s ", run_cmd)
+    status, output = process.getstatusoutput(run_cmd, timeout)
+
+    if output:
```
Comment (PaulYuuu) [inline qemu/tests/kvm_unit_test_nested.py:31]: Check `status` wold more meaningful?
`output` always exists.

File: qemu/tests/kvm_unit_test_nested.py (hunk 12 above)
Comment (qcheng-redhat) [reply qemu/tests/kvm_unit_test_nested.py]: Fixed.

File: qemu/tests/kvm_unit_test_nested.py (hunk 13 above)
Comment (qcheng-redhat) [reply qemu/tests/kvm_unit_test_nested.py:31]: The command filtered failed case and expected failures. So, if no failed test case, there will be no output.

File: qemu/tests/cfg/kvm_unit_test_nested.cfg (hunk 11 above)
Comment (qcheng-redhat) [reply qemu/tests/cfg/kvm_unit_test_nested.cfg]: Fixed.

File: qemu/tests/cfg/kvm_unit_test_nested.cfg (hunk 10 above)
Comment (qcheng-redhat) [reply qemu/tests/cfg/kvm_unit_test_nested.cfg:4]: The test suite will start a qemu process for the tests itself, and will handle l2 guests,  no need to start other guests.

File: qemu/tests/kvm_unit_test_nested.py (hunk 13 above)
Comment (PaulYuuu) [reply qemu/tests/kvm_unit_test_nested.py:31]: Okay, makes sense.

File: qemu/tests/cfg/kvm_unit_test_nested.cfg (hunk 10 above)
Comment (PaulYuuu) [reply qemu/tests/cfg/kvm_unit_test_nested.cfg:4]: If so, you may need to update the manual test steps.
@CongLi, I am not sure the usage from the x86 side, you guys can confirm this item together.

Comment (CongLi) [reply qemu/tests/cfg/kvm_unit_test_nested.cfg:4]: @PaulYuuu Yes, no need to boot guest, it's only check nested function in host.
@qcheng-redhat Please keep the automation case is same as the polarion case steps.

Comment (qcheng-redhat) [reply qemu/tests/cfg/kvm_unit_test_nested.cfg:4]: I have updated test case in Polarion. @PaulYuuu @CongLi 
https://polarion.engineering.redhat.com/polarion/#/project/RedHatEnterpriseLinux7/workitem?id=RHEL7-19680

Comment (qcheng-redhat) [reply qemu/tests/cfg/kvm_unit_test_nested.cfg]: Updated with starting 4 l1 guest vms on host.

File: qemu/tests/kvm_unit_test_nested.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    """
+    Run nested relstead tests in kvm-unit-test test suite
+
+    1) Start multiple(4) L1 guest vms
+    2) Clone kvm-unit-tests test suite from repo
+    3) Compile test suite
+    4) Run vmx/svm test suite with multiple running vms on host
+    """
+
+    params["start_vm"] = "yes"
```
Comment (PaulYuuu) [inline qemu/tests/kvm_unit_test_nested.py:19]: Meaningless line.

Comment (qcheng-redhat) [reply qemu/tests/kvm_unit_test_nested.py]: Removed.

**Comment by CongLi** [comment]
@PaulYuuu Could you please review this PR again?

Thanks.

**Review by PaulYuuu — APPROVED** [review]

**Comment by PaulYuuu** [comment]
Nothing else on my side.

**Comment by CongLi** [comment]
@PaulYuuu Could you please merge this patch? No need other review, thanks.

**Comment by PaulYuuu** [comment]
Ohhh, @qcheng-redhat please add `only x86` in cfg and provide the test result, thanks.

**Comment by qcheng-redhat** [comment]
python3 ConfigTest.py --testcase=kvm_unit_test_nested --guestname=RHEL.8.3.0 --machine=q35 --clone=no

 (1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.kvm_unit_test_nes\nFAIL: PDPTE2 is 0... (83.58 s)(expected 0x644001)3put: FAIL: Test Guest Segment Selector, GUEST_SEL_SS 13

The failure due to Bug 1881857 - vmx_mtf_pdpte_test test suite failed on Intel host

**Comment by PaulYuuu** [comment]
Please do not add the bug ID in the commit message.

**Comment by CongLi** [comment]
@PaulYuuu Thanks for the comment, could you check again?

**Pull request merged**